src/ray/gcs/gcs_server/gcs_placement_group_manager.cc

```diff
@@ -153,24 +172,122 @@ void GcsPlacementGroupManager::HandleCreatePlacementGroup(
+                    << " is removed successfully.";
+    } else {
+      RAY_LOG(WARNING) << "Removing a placement group of an id, " << placement_group_id
+                       << " failed.";
+    }
+    GCS_RPC_SEND_REPLY(send_reply_callback, reply, status);
+  });
+}
+
+void GcsPlacementGroupManager::RemovePlacementGroup(
+    const PlacementGroupID &placement_group_id,
+    StatusCallback on_placement_group_removed) {
+  RAY_CHECK(on_placement_group_removed);
+  // If the placement group is not registered, it means it is already removed or wasn't
+  // created before. Just fail the request in this case.
+  if (registered_placement_groups_.find(placement_group_id) ==
+      registered_placement_groups_.end()) {
+    std::stringstream stream;
+    stream << "Placement group of id " << placement_group_id
+           << " is already removed or doesn't exist";
+    on_placement_group_removed(Status::NotFound(stream.str()));
+    return;
+  }
+
+  // Remove the placement group if it is already created.
+  auto created_placement_group_it = created_placement_group_.find(placement_group_id);
+  if (created_placement_group_it != created_placement_group_.end()) {
+    gcs_placement_group_scheduler_->DestroyPlacementGroupBundleResources(
+        placement_group_id);
+    created_placement_group_.erase(created_placement_group_it);
+  }
+
+  // Cancel the scheduling requset if necessary.
+  if (IsSchedulingInProgress(placement_group_id)) {
+    // If the placement group is scheduling.
+    gcs_placement_group_scheduler_->MarkScheduleCancelled(placement_group_id);
+  }
+
+  // Remove a placement group from a pending list if exists.
+  auto pending_it = std::find_if(
+      pending_placement_groups_.begin(), pending_placement_groups_.end(),
+      [placement_group_id](const std::shared_ptr<GcsPlacementGroup> &placement_group) {
+        return placement_group->GetPlacementGroupID() == placement_group_id;
+      });
+  if (pending_it != pending_placement_groups_.end()) {
+    // The placement group was pending scheduling, remove it from the queue.
+    pending_placement_groups_.erase(pending_it);
+  }
+
+  // Remove registered placement group.
+  auto placement_grouop_it = registered_placement_groups_.find(placement_group_id);
+  auto placement_group = placement_grouop_it->second;
+  registered_placement_groups_.erase(placement_grouop_it);
+  // If placement group hasn't been created yet, send a response to a core worker that
+  // the creation of placement group has failed.
+  auto it = placement_group_to_register_callback_.find(placement_group_id);
+  if (it != placement_group_to_register_callback_.end()) {
+    std::stringstream stream;
+    stream << "Placement group of id " << placement_group_id
+           << " is removed before it is created";
+    it->second(Status::NotFound(stream.str()));
+    placement_group_to_register_callback_.erase(it);
+  }
+
+  placement_group->UpdateState(rpc::PlacementGroupTableData::REMOVED);
```

Comment:
@ffbin Do you think we should update all in-memory state after we persist the state to GCS? It seems like the current approach can have troubles when we start taking care of fault tolerance.

Reply:
PUT is asynchronous, if we update all in-memory state after we persist the state to GCS, there will be problems processing new requests before PUT is finished.